### PR TITLE
fixes to 32-bit handling, to support 32-bit arm

### PR DESCRIPTION
**Pull request by andrewrk**
With these changes, I was able to use ziglang/bootstrap to cross-compile an arm-linux-musleabihf tarball.

### CODE REVIEW

File: lib/std/zig/system.zig [hunk 1]
```diff
@@ -811,7 +815,7 @@ pub const NativeTargetInfo = struct {
     }
 
     fn preadMin(file: fs.File, buf: []u8, offset: u64, min_read_len: usize) !usize {
-        var i: u64 = 0;
+        var i: usize = 0;
```
Comment (daurnimator) [inline lib/std/zig/system.zig]: What if I have a large (>4GB) ELF file?

File: lib/std/zig/system.zig [hunk 2]
```diff
@@ -751,7 +752,10 @@ pub const NativeTargetInfo = struct {
                     const strtab_read_len = try preadMin(file, &strtab_buf, ds.offset, shstrtab_len);
                     const strtab = strtab_buf[0..strtab_read_len];
                     // TODO this pointer cast should not be necessary
-                    const rpath_list = mem.spanZ(@ptrCast([*:0]u8, strtab[rpoff..].ptr));
+                    const rpoff_usize = std.math.cast(usize, rpoff) catch |err| switch (err) {
+                        error.Overflow => return error.InvalidElfFile,
+                    };
+                    const rpath_list = mem.spanZ(@ptrCast([*:0]u8, strtab[rpoff_usize..].ptr));
```
Comment (daurnimator) [inline lib/std/zig/system.zig]: Is this was `strtab[rpoff_usize..:0].` could we remove the `@ptrCast`? (and TODO above)

File: lib/std/zig/system.zig (hunk 1 above)
Comment (andrewrk) [reply lib/std/zig/system.zig]: This fn is allowed to return short reads. `i` is just an offset, which is relative to `min_read_len`. `usize` was the correct type all along.

File: lib/std/zig/system.zig [hunk 3]
```diff
@@ -645,7 +646,7 @@ pub const NativeTargetInfo = struct {
                     {
                         var dyn_off = elfInt(is_64, need_bswap, ph32.p_offset, ph64.p_offset);
                         const p_filesz = elfInt(is_64, need_bswap, ph32.p_filesz, ph64.p_filesz);
-                        const dyn_size: u64 = if (is_64) @sizeOf(elf.Elf64_Dyn) else @sizeOf(elf.Elf32_Dyn);
+                        const dyn_size: usize = if (is_64) @sizeOf(elf.Elf64_Dyn) else @sizeOf(elf.Elf32_Dyn);
```
Comment (daurnimator) [inline lib/std/zig/system.zig]: Does peer-type resolution not work here? (i.e. omit the type?)

Comment (andrewrk) [reply lib/std/zig/system.zig]: `is_64` is runtime known, so it hits  #137

Comment (daurnimator) [reply lib/std/zig/system.zig]: oh interesting. I thought `@sizeOf` would return a comptime-known `usize` rather than a `comptime_int`

**Pull request merged**